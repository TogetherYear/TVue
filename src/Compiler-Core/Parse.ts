import { NodeType } from "./AST"

interface IContext {
    rawSource: string,
    source: string
}

interface IInterpolation {
    type: NodeType,
    content: {
        type: NodeType,
        content: string
    }
}

const enum Symbol {
    OpenDelimiter = '{{',
    CloaseDelimiter = '}}',
}

export const BaseParse = (content: string) => {
    const context = CreateParseContext(content)
    return CreateRoot(ParseChildren(context))
}

const CreateRoot = (children: Array<IInterpolation>) => {
    return {
        children
    }
}

const ParseChildren = (context: IContext) => {
    const nodes = []
    if (context.source.startsWith(Symbol.OpenDelimiter)) {
        const node = ParseInterpolation(context)
        nodes.push(node)
    }
    return nodes
}

const ParseInterpolation = (context: IContext): IInterpolation => {
    const closeIndex = context.source.indexOf(Symbol.CloaseDelimiter, Symbol.OpenDelimiter.length)
    const content = context.source.slice(Symbol.OpenDelimiter.length, closeIndex)
    context.source = context.source.slice(closeIndex + Symbol.CloaseDelimiter.length)
    return {
        type: NodeType.Interpolation,
        content: {
            type: NodeType.SimpleExpression,
            content: content,
        }
    }
}

const CreateParseContext = (content: string) => {
    const context = {
        rawSource: content,
        source: content.replace(/[\n\t\s]+/g, '')
    } as IContext
    return context
}