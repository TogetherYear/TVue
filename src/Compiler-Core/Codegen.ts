import { NodeType } from "./AST";
import { INode, INodeContent } from "./Parse";
import { RuntimeHelper } from "./RuntimeHelper";

interface ICodegenContext {
    code: string,
    Push: (source: string) => void
}

export const Generate = (root: INode) => {
    const context = CreateCodegenContext()
    const { Push } = context

    GenFunctionPreamble(root, context)

    const funcName = 'Render'
    const args = ['_ctx,_cache']
    const signature = args.join(', ')

    Push(`function ${funcName}(${signature}) {`)
    Push('return ')
    GenNode(root.codegenNode!, context)
    Push('}')

    return {
        code: context.code
    }
}

const GenFunctionPreamble = (root: INode, context: ICodegenContext) => {
    const { Push } = context
    const vueBinging = `TVue`
    if (root.helpers && root.helpers.length != 0) {
        Push(`const { ${(root.helpers || []).map(h => `${h}:_${h}`).join(',')} } = ${vueBinging}`)
        Push(`\n`)
    }
    Push('return ')
}

const CreateCodegenContext = () => {
    const context = {
        code: '',
        Push: (source: string) => {
            context.code += source
        }
    } as ICodegenContext
    return context
}

const GenNode = (node: INode | INodeContent, context: ICodegenContext) => {
    switch (node.type) {
        case NodeType.Text:
            GenText(node, context)
            break;
        case NodeType.Interpolation:
            GenInterpolation(node, context)
            break;
        case NodeType.SimpleExpression:
            GenExpression(node, context)
        case NodeType.Element:
            GenElement(node, context)
    }
}

const GenText = (node: INode, context: ICodegenContext) => {
    const { Push } = context
    Push(`'${node.content}'`)
}

const GenInterpolation = (node: INode, context: ICodegenContext) => {
    const { Push } = context
    Push(`_${RuntimeHelper.ToDisplayString}(`)
    GenNode(node.content as INodeContent, context)
    Push(`)`)
}

const GenExpression = (node: INode, context: ICodegenContext) => {
    const { Push } = context
    Push(`${node.content}`)
}

const GenElement = (node: INode, context: ICodegenContext) => {
    const { Push } = context
    Push(`_${RuntimeHelper.CreateElementVNode}('div')`)
}
