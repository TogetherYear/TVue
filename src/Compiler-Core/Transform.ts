import { NodeType } from "./AST";
import { INode } from "./Parse";

export interface ITransformOptions {
    nodeTransforms: Array<(node: INode) => void>
}

interface ITransformContext {
    root: INode,
    nodeTransforms: Array<(node: INode) => void>
}

export const Transform = (root: INode, options?: ITransformOptions) => {
    const context = CreateTransformContext(root, options)
    TraverseNode(root, context)
}

const CreateTransformContext = (root: INode, options?: ITransformOptions) => {
    const context = {
        root,
        nodeTransforms: options?.nodeTransforms || []
    } as ITransformContext
    return context
}

const TraverseNode = (node: INode, context: ITransformContext) => {
    const nodeTransforms = context.nodeTransforms
    for (let i = 0; i < nodeTransforms.length; ++i) {
        const Tf = nodeTransforms[i]
        Tf(node)
    }

    TransformChildren(node, context)
}

const TransformChildren = (node: INode, context: ITransformContext) => {
    const children = node.children
    if (children) {
        for (let i = 0; i < children.length; ++i) {
            const node = children[i]
            TraverseNode(node, context)
        }
    }
}