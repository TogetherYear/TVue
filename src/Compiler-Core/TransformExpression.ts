import { NodeType } from "./AST";
import { INode, INodeContent } from "./Parse";

export const TransformExpression = (node: INode) => {
    if (node.type === NodeType.Interpolation) {
        const originContent = node.content as INodeContent;
        const rawContent = originContent.content;
        originContent.content = `_ctx.${rawContent}`
    }
}