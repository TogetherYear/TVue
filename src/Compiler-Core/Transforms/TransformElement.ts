import { NodeType } from "../AST";
import { INode } from "../Parse";
import { RuntimeHelper } from "../RuntimeHelper";
import { ITransformContext } from "../Transform";

export const TransformElement = (node: INode, context: ITransformContext) => {
    if (node.type === NodeType.Element) {
        context.Helper(RuntimeHelper.CreateElementVNode)
    }
}