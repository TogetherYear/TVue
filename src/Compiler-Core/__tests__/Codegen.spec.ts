import { Generate } from "../Codegen"
import { BaseParse } from "../Parse"
import { Transform } from "../Transform"
import { TransformElement } from "../Transforms/TransformElement"
import { TransformExpression } from "../Transforms/TransformExpression"

describe('codegen', () => {
    test('string', () => {
        const str = `TVue`
        const ast = BaseParse(str)
        Transform(ast)
        const code = Generate(ast)
        expect(code).toMatchSnapshot()
    })

    test('interpolation', () => {
        const str = `{{ message }}`
        const ast = BaseParse(str)
        Transform(ast, {
            nodeTransforms: [TransformExpression]
        })
        const code = Generate(ast)
        expect(code).toMatchSnapshot()
    })

    // test("element", () => {
    //     const str = `<div>hi,{{message}}</div>`
    //     const ast: any = BaseParse(str);
    //     Transform(ast, {
    //         nodeTransforms: [TransformElement],
    //     });

    //     const { code } = Generate(ast);
    //     expect(code).toMatchSnapshot();
    // });
})