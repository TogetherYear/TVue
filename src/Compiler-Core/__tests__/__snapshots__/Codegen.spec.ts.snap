// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { ToDisplayString:_ToDisplayString } = TVue
return function Render(_ctx,_cache) {return _CreateElementVNode('div')}"
`;

exports[`codegen interpolation 1`] = `
{
  "code": "const { ToDisplayString:_ToDisplayString } = TVue
return function Render(_ctx,_cache) {return _ToDisplayString(_ctx.message_CreateElementVNode('div'))}",
}
`;

exports[`codegen string 1`] = `
{
  "code": "return function Render(_ctx,_cache) {return 'TVue'}",
}
`;
